src/backend/distributed/planner/shard_pruning.c

```diff
@@ -547,33 +549,93 @@ PrunableExpressionsWalker(Node *node, ClauseWalkerContext *context)
 		Node *strippedLeftOpExpression = strip_implicit_coercions(leftOpExpression);
 		bool usingEqualityOperator = OperatorImplementsEquality(
 			arrayOperatorExpression->opno);
+		Expr *arrayArgument = (Expr *) lsecond(arrayOperatorExpression->args);
 
 		/*
-		 * Citus cannot prune hash-distributed shards with ANY/ALL. We show a NOTICE
-		 * if the expression is ANY/ALL performed on the partition column with equality.
-		 *
-		 * TODO: this'd now be easy to implement, similar to the OR_EXPR case
-		 * above, except that one would push an appropriately constructed
-		 * OpExpr(LHS = $array_element) as continueAt.
+		 * Found partcol = ANY(const, value, s); or partcol IN (const,b,c);
 		 */
 		if (usingEqualityOperator && strippedLeftOpExpression != NULL &&
-			equal(strippedLeftOpExpression, context->partitionColumn))
+			equal(strippedLeftOpExpression, context->partitionColumn) &&
+			IsA(arrayArgument, Const))
 		{
-			ereport(NOTICE, (errmsg("cannot use shard pruning with "
-									"ANY/ALL (array expression)"),
-							 errhint("Consider rewriting the expression with "
-									 "OR/AND clauses.")));
-		}
+			ArrayType *array = NULL;
+			int16 typlen = 0;
+			bool typbyval = false;
+			char typalign = '\0';
+			Oid element_type = 0;
+			ArrayIterator arrayIterator = NULL;
+			Datum arg = 0;
+			Datum arrayDatum = ((Const *) arrayArgument)->constvalue;
+			bool isNull = false;
+
+			/* check for the NULL righthand expression*/
+			if (arrayDatum == 0)
+			{
+				return false;
+			}
 
-		/*
-		 * Mark expression as added, so we'll fail pruning if there's no ANDed
-		 * restrictions that we can deal with.
-		 */
-		if (!prune->addedToPruningInstances)
+			array = DatumGetArrayTypeP(((Const *) arrayArgument)->constvalue);
+
+			/* get the necessary information from array type to iterate over it */
+			element_type = ARR_ELEMTYPE(array);
+			get_typlenbyvalalign(element_type,
+								 &typlen,
+								 &typbyval,
+								 &typalign);
+
+			/* Iterate over the righthand array of expression */
+			arrayIterator = array_create_iterator(array, 0, NULL);
+			while (array_iterate(arrayIterator, &arg, &isNull))
+			{
+				OpExpr *op = NULL;
+				PendingPruningInstance *instance =
+					palloc0(sizeof(PendingPruningInstance));
+				Const *c = NULL;
+
+				/* build partcol = arrayelem operator */
+				op = makeNode(OpExpr);
+				op->opno = arrayOperatorExpression->opno;
+				op->opfuncid = arrayOperatorExpression->opfuncid;
+				op->inputcollid = arrayOperatorExpression->inputcollid;
+				op->opresulttype = get_func_rettype(arrayOperatorExpression->opfuncid);
+				op->opcollid = DEFAULT_COLLATION_OID;
+				op->location = -1;
+
+				c = makeConst(element_type, -1,
```

Comment:
Why not create the `Const` first, directly when defining the variable?  Then creation of the `OpExpr` isn't separated from filling out `op->args` as well.